File: data_structures_by_python/linked_list/singly_linked_list/delete_at_tail.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

def insert_at_tail(head, tail, val):
    new_node = Node(val)
    if head is None:
        head = new_node
        tail = new_node
        return head, tail
    tail.next = new_node
    tail = new_node
    return head, tail

def delete_at_tail(head, tail):
    if head is None:
        return head, tail
    if head.next is None:
        head = None
        tail = None
        return head, tail
    temp = head
    while temp.next.next is not None:
        temp = temp.next
    temp.next = None
    tail = temp
    return head, tail

def print_linked_list(head):
    temp = head
    while temp is not None:
        print(temp.val, end=" ")
        temp = temp.next
    print()

head = None
tail = None
head, tail = insert_at_tail(head, tail, 10)
head, tail = insert_at_tail(head, tail, 20)
head, tail = insert_at_tail(head, tail, 30)
head, tail = insert_at_tail(head, tail, 40)

print_linked_list(head)
print("Tail before delete:", tail.val)

head, tail = delete_at_tail(head, tail)

print_linked_list(head)
print("Tail after delete:", tail.val)

File: data_structures_by_c++/linked_list/singly_linked_list/delete_at_tail.cpp
#include <bits/stdc++.h>
using namespace std;

class Node
{
public:
    int val;
    Node* next;

    Node(int val)
    {
        this->val = val;
        this->next = NULL;
    }
};

void insert_at_tail(Node* &head, Node* &tail, int val)
{
    Node* newnode = new Node(val);
    if(head == NULL)
    {
        head = newnode;
        tail = newnode;
        return;
    }
    tail->next = newnode;
    tail = newnode;
}

void delete_at_tail(Node* &head, Node* &tail)
{
    if(head == NULL) return;
    if(head->next == NULL)
    {
        delete head;
        head = NULL;
        tail = NULL;
        return;
    }
    Node* tmp = head;
    while(tmp->next->next != NULL)
    {
        tmp = tmp->next;
    }
    delete tmp->next;
    tmp->next = NULL;
    tail = tmp;
}

void print_linked_list(Node* head)
{
    Node* tmp = head;
    while(tmp != NULL)
    {
        cout << tmp->val << " ";
        tmp = tmp->next;
    }
    cout << endl;
}

int main()
{
    Node* head = NULL;
    Node* tail = NULL;

    insert_at_tail(head, tail, 10);
    insert_at_tail(head, tail, 20);
    insert_at_tail(head, tail, 30);
    insert_at_tail(head, tail, 40);

    print_linked_list(head);
    cout << "Tail before delete : " << tail->val << endl;
    delete_at_tail(head,tail);
    print_linked_list(head);
    cout << "Tail after delete: " << tail->val << endl;
    return 0;
}
